hazelcast/src/main/java/com/hazelcast/cp/event/impl/CPGroupAvailabilityEventGracefulImpl.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.cp.event.impl;
+
+import com.hazelcast.cp.CPGroup;
+import com.hazelcast.cp.CPGroupId;
+import com.hazelcast.cp.CPMember;
+import com.hazelcast.cp.event.CPGroupAvailabilityEvent;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Objects;
+
+import static com.hazelcast.internal.serialization.impl.SerializationUtil.readCollection;
+import static com.hazelcast.internal.serialization.impl.SerializationUtil.writeCollection;
+
+// it's probably easier to just add a flag to the existing availability message, e.g. isShutdown -- default value of false
+public class CPGroupAvailabilityEventGracefulImpl implements CPGroupAvailabilityEvent, IdentifiedDataSerializable {
```

Comment:
Can we subclass
`CPGroupAvailabilityEventGracefulImpl extends CPGroupAvailabilityEventImpl` ?